### ISSUE TITLE
Core dump on exit (Python 3.5, OSX)

### ISSUE THREAD

**Issue by mhils** [pyca/cryptography#3207]
We are currently observing cryptography/pyOpenSSL to cause core dumps on exit with Python 3.5:

```
illegal hardware instruction (core dumped) mitmproxy

* thread #1: tid = 0x0000, 0x0000000110402f60, stop reason = signal SIGSTOP
  * frame #0: 0x0000000110402f60
    frame #1: 0x00000001101fa5a2 _openssl.abi3.so`SSL_free + 49
    frame #2: 0x000000011024b79a _openssl.abi3.so`_cffi_f_SSL_free + 138
    frame #3: 0x000000010f077065 .Python`PyCFunction_Call + 276
    frame #4: 0x000000010f03ed46 .Python`PyObject_Call + 101
    frame #5: 0x000000010f03be65 .Python`PyObject_CallFunctionObjArgs + 190
    frame #6: 0x00000001103e7d56 _cffi_backend.cpython-35m-darwin.so`gcp_finalize + 54
    frame #7: 0x000000010f06eda8 .Python`dict_dealloc + 120
    frame #8: 0x000000010f0847c7 .Python`subtype_dealloc + 662
    frame #9: 0x000000010f06eda8 .Python`dict_dealloc + 120
    frame #10: 0x000000010f0847c7 .Python`subtype_dealloc + 662
    frame #11: 0x000000010f06eda8 .Python`dict_dealloc + 120
    frame #12: 0x000000010f08ced3 .Python`subtype_clear + 145
    frame #13: 0x000000010f10e1eb .Python`collect + 1855
    frame #14: 0x000000010f10daa3 .Python`_PyGC_CollectNoFail + 43
    frame #15: 0x000000010f0edb90 .Python`PyImport_Cleanup + 982
    frame #16: 0x000000010f0f784c .Python`Py_Finalize + 98
    frame #17: 0x000000010f0f7dd7 .Python`Py_Exit + 13
    frame #18: 0x000000010f0fa34a .Python`handle_system_exit + 318
    frame #19: 0x000000010f0f9fc0 .Python`PyErr_PrintEx + 41
    frame #20: 0x000000010f0f9659 .Python`PyRun_SimpleFileExFlags + 908
    frame #21: 0x000000010f10d36d .Python`Py_Main + 3301
    frame #22: 0x000000010f030e3f python3.5`___lldb_unnamed_function1$$python3.5 + 224
    frame #23: 0x00007fffb9297255 libdyld.dylib`start + 1
```

Unfortunately we don't have a good way to reproduce this issue yet, but this is likely related to #2970 and we wanted to give you a quick heads-up. Any pointers to why this is happening or how we could fix this would be highly appreciated. The mitmproxy issue for this with a few more details is https://github.com/mitmproxy/mitmproxy/issues/1645! 😃

**Comment by reaperhulk** [comment]
Python 3.5 interpreter shutdown is the worst...

**Comment by reaperhulk** [comment]
@mhils can you see if this is still occurring in current master? We've completely replaced our locking callbacks with a C-based approach so *fingers crossed*.

**Comment by reaperhulk** [comment]
This should have been resolved with the release of 1.6 (which removed our python-based locking callbacks).
